src/Microsoft.TemplateEngine.Core/Expressions/MSBuild/MSBuildStyleEvaluatorDefinition.cs

```diff
@@ -0,0 +1,397 @@
+                        rd = rh;
+                    }
+                    else
+                    {
+                        return null;
+                    }
+                }
+            }
+
+            return ld.CompareTo(rd);
+        }
+
+        private static int? AttemptVersionComparison(object left, object right)
+        {
+            Version lv = left as Version;
+
+            if (lv == null)
+            {
+                string ls = left as string;
+                if (ls == null || !Version.TryParse(ls, out lv))
+                {
+                    return null;
+                }
+            }
+
+            Version rv = right as Version;
+
+            if (rv == null)
+            {
+                string rs = right as string;
+                if (rs == null || !Version.TryParse(rs, out rv))
+                {
+                    return null;
+                }
+            }
+
+            return lv.CompareTo(rv);
+        }
+
+        private static int Compare(object left, object right)
+        {
+            return AttemptNumericComparison(left, right)
+                   ?? AttemptVersionComparison(left, right)
+                   ?? AttemptLexographicComparison(left, right)
+                   ?? AttemptComparableComparison(left, right)
+                   ?? 0;
+        }
+
+        private static ITokenTrie GetSymbols(IProcessorState processor)
+        {
+            ITokenTrie tokens;
+            if (!TokenCache.TryGetValue(processor.Encoding, out tokens))
+            {
+                TokenTrie trie = new TokenTrie();
+
+                //Logic
+                trie.AddToken(processor.Encoding.GetBytes("AND"));
+                trie.AddToken(processor.Encoding.GetBytes("OR"));
+                trie.AddToken(processor.Encoding.GetBytes("!"));
+                trie.AddToken(processor.Encoding.GetBytes("&gt;"));
+                trie.AddToken(processor.Encoding.GetBytes("&gt;="));
+                trie.AddToken(processor.Encoding.GetBytes("&lt;"));
+                trie.AddToken(processor.Encoding.GetBytes("&lt;="));
+                trie.AddToken(processor.Encoding.GetBytes("=="));
+                trie.AddToken(processor.Encoding.GetBytes("!="));
+
+                //Braces
+                trie.AddToken(processor.Encoding.GetBytes("("));
+                trie.AddToken(processor.Encoding.GetBytes(")"));
+
+                //Whitespace
+                trie.AddToken(processor.Encoding.GetBytes(" "));
+                trie.AddToken(processor.Encoding.GetBytes("\t"));
+
+                //EOLs
+                trie.AddToken(processor.Encoding.GetBytes("\r\n"));
+                trie.AddToken(processor.Encoding.GetBytes("\n"));
+                trie.AddToken(processor.Encoding.GetBytes("\r"));
+
+                // quotes
+                trie.AddToken(processor.Encoding.GetBytes("'"));
+
+                // variable start
+                trie.AddToken(processor.Encoding.GetBytes("$("));
```

Comment:
Not in love with this, looking for ideas (please comment). This works as follows:

1) All variables in this document are known to be of the form $(FOO)
2) All variables are part of the trie and therefore will be consumed as a whole unit
3) In the event that a free $( is found, it's not part of a known symbol (binds to the BadSyntax token type)
4) The BadSyntax token is checked for in the scope builder and rejects the expression if encountered
5) If the builder rejects the expression, all of the bytes for the expression are written back to the target (via a callback which presumably also emits anything else that has been excluded from the target stream pending the result of the evaluation)